source/sema.h

```diff
@@ -19,23 +19,147 @@
 #define CPP2_SEMA_H
 
 #include "reflect.h"
+#include "dll.h"
 
 
 namespace cpp2 {
 
+static reflect_libraries* active_instance = nullptr;
+
+class reflect_libraries
+{
+    std::vector<error_entry>& errors;
+    std::string               in_progress = {};
+    std::vector<dll>          loaded      = {};
+
+    struct function_entry
+    {
+        std::variant<
+            std::monostate,
+            meta::mf_sign_type_in*,
+            meta::mf_sign_type_inout*
+        > type = {};
+        std::variant<
+            std::monostate,
+            meta::mf_sign_func_in*,
+            meta::mf_sign_func_inout*
+        > func = {};
+    };
+
+    std::map<std::string, function_entry> functions = {};
+
+public:
+    reflect_libraries(std::vector<error_entry>& errors_)
+        : errors{ errors_ }
+    { }
+
+    void load(std::string_view lib) {
+        active_instance = this;
+        in_progress = lib;
+        finally _{ [&](){ active_instance = nullptr; in_progress = ""; } };
+        auto& dl = loaded.emplace_back(in_progress);
+        if(!dl.is_open()) {
+            errors.emplace_back(
+                source_position(-1, -1),
+                "unable to load library '" + in_progress + "': " + dll::get_last_error()
+            );
+        }
+    }
+
+    bool try_apply(
+        [[maybe_unused]] std::string_view caller,
+        std::string const& name,
+        auto& decl
+    ) const {
+        // TODO(DyXel): Do full/better name look-up.
+        auto search = functions.find(name);
+        if(search == functions.end()) return false;
+        auto& entry = search->second;
+        auto visitor = [&](auto&& v) -> bool {
+            if constexpr(requires { v(decl); }) {
+                v(decl);
+                return true;
+            }
+            return false;
+        };
+        if constexpr(std::is_same_v<meta::type_declaration, CPP2_TYPEOF(decl)>)
+            return std::visit(visitor, entry.type);
+        else
+            return std::visit(visitor, entry.func);
+    }
+
+private:
+    friend meta::register_metafunction;
+
+    void register_one(std::string name, auto* func) {
+        // TODO(DyXel): Check if already registered
+        auto& entry = functions[name];
+        if constexpr(requires { entry.type = func; })
```

Comment:
I find this `requires { entry.type = func; }` hard to read.

Reply:
Is it? this code just means "if we can assign to the `type` variant, assign to it, otherwise assign to the `func` variant. Again, the alternative would be a complicated lambda overload trick mess to assign the value to the correct variant. Same observation as above applies, this is all throwaway code.

Reply:
With respect to the lambda it is probably simpler to read. I might just need to get used to it.